packages/functions/src/vertex-paint.ts

```diff
@@ -0,0 +1,152 @@
+export interface VertexPaintOptions {}
+
+const VERTEX_PAINT_DEFAULTS: VertexPaintOptions = {};
+
+/**
+ * Paints each {@link Material}'s base color texture to vertex colors, and removes the texture.
+ *
+ * Example:
+ *
+ * ```javascript
+ * import { vertexPaint } from '@gltf-transform/functions';
+ *
+ * const mesh = document.getRoot().listMeshes()
+ * 	.find((mesh) => mesh.getName() === 'MyMesh');
+ * const prim = mesh.listPrimitives()[0];
+ *
+ * prim.getAttribute('COLOR_0'); // → null
+ *
+ * await document.transform(vertexPaint());
+ *
+ * prim.getAttribute('COLOR_0'); // → Accessor
+ * ```
+ */
+export function vertexPaint(_options: VertexPaintOptions = VERTEX_PAINT_DEFAULTS): Transform {
+	// eslint-disable-next-line @typescript-eslint/no-unused-vars
+	const options = { ...VERTEX_PAINT_DEFAULTS, ..._options } as Required<VertexPaintOptions>;
+
+	return createTransform(NAME, async (document: Document): Promise<void> => {
+		const root = document.getRoot();
+		const logger = document.getLogger();
+
+		const pixelsLookup = new Map<Material, { pixels: NdArray<Uint8Array>; texCoord: number }>();
+
+		// Pre-process; load pixel data into memory.
+		for (const material of root.listMaterials()) {
+			const texture = material.getBaseColorTexture();
+			const textureInfo = material.getBaseColorTextureInfo();
+			if (!texture || !textureInfo) continue;
+
+			const pixels = (await getPixels(texture.getImage()!, texture.getMimeType())) as NdArray<Uint8Array>;
+			const texCoord = textureInfo.getTexCoord();
+			pixelsLookup.set(material, { pixels, texCoord });
+			material.setBaseColorTexture(null);
+		}
+
+		// For each mesh primitive, paint vertex colors.
+		for (const mesh of root.listMeshes()) {
+			for (const prim of mesh.listPrimitives()) {
+				const material = prim.getMaterial();
+				if (!material || !pixelsLookup.has(material)) continue;
+
+				const { pixels, texCoord } = pixelsLookup.get(material)!;
+				_vertexPaintPrimitive(document, prim, pixels, texCoord);
+			}
+		}
+
+		// Clean up unused textures and texture coordinates.
+		for (const [texture] of pixelsLookup) texture.dispose();
+		await document.transform(prune({ propertyTypes: [PropertyType.ACCESSOR] }));
+
+		logger.debug(`${NAME}: Complete.`);
+	});
+}
+
+function _vertexPaintPrimitive(
+	document: Document,
+	prim: Primitive,
+	pixels: NdArray<Uint8Array>,
+	texCoord: number
+): void {
+	const uv = prim.getAttribute(`TEXCOORD_${texCoord}`)!;
+	const color = document
+		.createAccessor()
+		.setArray(new Float32Array(uv.getCount() * 3))
+		.setBuffer(uv.getBuffer())
+		.setType('VEC3'); // TODO(bug): Consider opacity?
+
+	const _uv = [0, 0] as vec2;
+	const _color = [0, 0, 0] as vec3;
+
+	// Read color from texture, write to COLOR_0 vertex attribute.
+	for (let i = 0, il = uv.getCount(); i < il; i++) {
+		uv.getElement(i, _uv);
+		const x = Math.floor(_uv[0] * pixels.shape[0]);
+		const y = Math.floor(_uv[1] * pixels.shape[1]);
```

Comment:
Could `bilinear` and `bicubic` be optional used here?

Reply:
Possibly — I think adding bilinear interpolation would be easy enough at least. For use cases I have in mind, the interpolation probably won't matter very much. If you think this affects your models, test cases would be welcome!